ibis/backends/__init__.py

```diff
@@ -1110,13 +1111,11 @@ def _register_udfs(self, expr: ir.Expr) -> None:
         if self.supports_python_udfs:
             raise NotImplementedError(self.name)
 
-    def _in_memory_table_exists(self, name: str) -> bool:
-        return name in self.list_tables()
-
     def _register_in_memory_tables(self, expr: ir.Expr) -> None:
         for memtable in expr.op().find(ops.InMemoryTable):
-            if not self._in_memory_table_exists(memtable.name):
+            if memtable not in self._memtables:
```

Comment:
I _think_ with this logic, we have the following fail case:

```python
t1 = ibis.memtable(..., name="test")
t2 = ibis.memtable(..., name="test")

# Use t1 on `con`, it's registered as `test`
con.execute(t1)

# Use t2 on `con`, it's registered, overwriting table `test`
con.execute(t2)

# Drop t1, its finalizer runs, deleting table `test`
del t1

# Try to execute against t2, it's still in the weakset so it's not re-registered, failure
con.execute(t2)
```

Reply:
This should be addresed now.